source/FormationPattern.cpp

```diff
@@ -91,9 +195,227 @@ void FormationPattern::SetName(const std::string &name)
+		positions += repeat->repeatPositions * ring;
+	}
+
+	// Compensate for any skipped positions. This would usually be the start-position on the line, but it can be the
+	// end position if we are on an alternating line.
+	if(ring % 2 && repeat && repeat->alternating && line.skipLast)
+		++linePosition;
+	else if(line.skipFirst)
+		++linePosition;
+
+	// Switch to arc-specific calculations if this line is an arc.
+	if(line.isArc)
+	{
+		// Calculate angles and radius from anchor to start.
+		double startAngle = Angle(startPx).Degrees();
+		double endAngle = line.angle;
+		double radius = startPx.Length();
+
+		// Apply repeat section for endAngle, startAngle and anchor were already done before.
+		if(repeat)
+		{
+			endAngle += repeat->repeatAngle * ring;
+
+			// Calculate new start and turn end angle if we need to alternate this repeat.
+			if(ring % 2 && repeat->alternating && positions > 0)
+			{
+				startAngle += endAngle;
+				endAngle = -endAngle;
+			}
+		}
+
+		// Apply positions to get the correct position-angle.
+		if(positions > 1)
+			endAngle /= positions - 1;
+		double positionAngle = startAngle + endAngle * linePosition;
+
+		// Get into the range of 0 to 360 for conversion to angle)
+		if(positionAngle < 0)
+			positionAngle = -fmod(-positionAngle, 360) + 360;
+		else
+			positionAngle = fmod(positionAngle, 360);
+
+		// Combine anchor with the position and return the result.
+		return endOrAnchorPx + Angle(positionAngle).Unit() * radius;
+	}
+
+	// This is not an arc, perform the line-based calculations.
+
+	// Swap start and end if we need to alternate in the repeat section.
+	// Apply repeat section if it is relevant.
+	if(ring % 2 && repeat && repeat->alternating)
+	{
+		Point tmpPx = endOrAnchorPx;
+		endOrAnchorPx = startPx;
+		startPx = tmpPx;
+	}
+
+	// Calculate the step from each position between start and end.
+	Point positionPx = endOrAnchorPx - startPx;
+
+	// Divide by positions, but don't count the first (since it is at position 0, not at position 1).
+	if(positions > 1)
+		positionPx /= positions - 1;
+
+	// Calculate position of the current position.
+	return startPx + positionPx * linePosition;
+}
+
+
+
+int FormationPattern::Rotatable() const
+{
+	return rotatable;
+}
+
+
+
+bool FormationPattern::FlippableY() const
+{
+	return flippable_y;
+}
+
+
+
+bool FormationPattern::FlippableX() const
+{
+	return flippable_x;
+}
+
+
+void FormationPattern::MultiAxisPoint::Add(Axis axis, const Point& toAdd)
+{
+	position[axis] += toAdd;
+}
+
+
+
+void FormationPattern::MultiAxisPoint::AddLoad(const DataNode &node)
+{
+	// We need at least the position keyword and 2 coordinate numbers.
+	if(node.Size() < 3)
+		return;
+
+	// By default we parse for pixels.
+	Axis axis = PIXELS;
+	double scalingFactor = 1.;
+
+	// Parse all the keywords before the coordinate
+	for(int i = 1; i < node.Size() - 2; ++i)
+		node.PrintTrace("Ignoring unrecognized token " + node.Token(i) + ":");
```

Comment:
why are you doing this?

Reply:
PR https://github.com/endless-sky/endless-sky/pull/4471 can parse many more tokens at this position than that this PR can. Quite a number of data-files already exist and I prefer to give an explicit warning if somebody tries to use an unsupported file, than to just silently ignore the extra tokens.

Reply:
ah, ok